aio: fix running against local dist builds of Angular

petebacondarwin opened the pull request:
Closes #19388
Closes #19411

petebacondarwin committed:
fix(aio): fix SearchService to work with TypeScript 2.4

The call to `race` required a type parameter to disambiguate the return type.

petebacondarwin committed:
build(aio): upgrade `ts-node` to support newer TypeScript config

This was causing `Cannot find type definition file for 'jasmine'. (2688)`
errors when running Protractor.

mary-poppins commented:
You can preview b587804 at https://pr19511-b587804.ngbuilds.io/.

mary-poppins commented:
You can preview ac3f525 at https://pr19511-ac3f525.ngbuilds.io/.

petebacondarwin commented:
Subsequent to landing this PR we will need to re-enable the aot-compiler example (https://github.com/angular/angular/issues/19411), which this PR doesn't actually do.

mary-poppins commented:
You can preview 77510fa at https://pr19511-77510fa.ngbuilds.io/.

petebacondarwin committed:
docs(aio): fix pipes example e2e spec to work with Angular v4

There was a breaking change between 4 and 5 for i18n formatting of dates.

petebacondarwin committed:
build(aio): ensure webdriver is updated when switching between local and npm deps

tbosch commented on aio/content/examples/aot-compiler/tsconfig-aot.json:
```diff
@@ -20,7 +20,8 @@
   "angularCompilerOptions": {
-   "genDir": "aot",
-   "skipMetadataEmit" : true
+    "alwaysCompileGeneratedCode": true,
```
Why do you need this? This will generate a `.ngfactory.js` file for all `.ts` files, even if it is empty. We only use it inside of G3 so far...

petebacondarwin replied:
I copied this from config in the `compiler-cli/integrationtest` that folder I think. I was just fishing for solutions :-)

tbosch commented on aio/content/examples/aot-compiler/tsconfig-aot.json:
```diff
@@ -20,7 +20,8 @@
-   "genDir": "aot",
-   "skipMetadataEmit" : true
+    "alwaysCompileGeneratedCode": true,
+    "genDir": ".",
```
You don't need `genDir` anymore.

petebacondarwin replied:
Unfortunately we need to leave this in to keep the examples compatible with Angular v4, which is what they are run against when authors are doing their "normal" work.

IgorMinar replied:
why don't we run the examples on the master against the recent v5 RC? then we don't need to worry about this.

petebacondarwin replied:
I was assuming that we would switch over to v5 once we released 5.0.0 proper. Then we can get rid of this (and the other hack in the animations example).

tbosch commented on aio/content/examples/.gitignore:
```diff
@@ -57,9 +57,8 @@ dist/
-!*/aot/bs-config.json
-!*/aot/index.html
+**/*.ngsummary.json
+**/*.factory.json
```
There are no `.factory.json` files (at least not from `ngc`)

petebacondarwin replied:
Right, I think that was an error :-)

petebacondarwin commented on aio/content/examples/animations/e2e/app.e2e-spec.ts:
```diff
@@ -335,17 +335,17 @@ describe('Animation Tests', () => {
-  function getOffsetWidth(el: ElementFinder): promise.Promise<number> {
-    return browser.executeScript(
+  function getOffsetWidth(el: ElementFinder) {
+    return browser.executeScript<number>(
```
I don't think I need this type param any more...

mary-poppins commented:
You can preview 11d0a2b at https://pr19511-11d0a2b.ngbuilds.io/.

petebacondarwin committed:
docs(aio): fix animations example to work with Angular v5

petebacondarwin committed:
build(aio): ignore @angular/mobile-toolkit in dist builds

The new v5 version of this toolkit is too incompatible for us to run
AIO against both versions. So for now we will ignore this libary when
running against local Angular distributions and continue to use the
version on npm.

petebacondarwin committed:
docs(aio): fix aot-compiler example to work with Angular v5

mary-poppins commented:
You can preview 6b74d4d at https://pr19511-6b74d4d.ngbuilds.io/.

mary-poppins commented:
You can preview ec21b7f at https://pr19511-ec21b7f.ngbuilds.io/.

gkalpak commented on aio/tools/ng-packages-installer/index.js:
```diff
@@ -0,0 +1,226 @@
+    .option('force', { describe: 'Force the command to execute even if not needed.', default: false })
+    .option('ignore-packages', { describe: 'List of Angular packages that should not be used in local mode.', default: [], array: true })
+
+    .command('overwrite <projectDir> [--force] [--debug] [--ignore-package package1 package2]', 'Install dependencies from the locally built Angular distributables.', () => {}, argv => {
```
_--ignore-package_ --> _--ignore-packages_

gkalpak commented on aio/tools/ng-packages-installer/index.js:
```diff
@@ -0,0 +1,226 @@
+      const installer = new NgPackagesInstaller(argv.projectDir, argv);
+      installer.installLocalDependencies();
+    })
+    .command('restore <projectDir> [--force] [--debug]', 'Install dependencies from the npm registry.', () => {}, argv => {
```
The `--force` option doesn't have any effect here afaict.

petebacondarwin replied:
That is true!

gkalpak replied:
We have tests to catch regressions in code while refactoring.
But we need a tool to catch regressions in docs :grin:

gkalpak reviewed (commented):
A few minor suggestions/questions, but overall looks great 👍
(Travis agrees, so not much to argue there 😛)

💯 on figuring this all out. By looking at the commits, it must have been quite an adventure 🚵

petebacondarwin commented on aio/package.json:
```diff
@@ -6,6 +6,9 @@
   "author": "Angular",
   "license": "MIT",
   "scripts": {
+    "aio-use-local": "node tools/ng-packages-installer overwrite . --debug --ignore-packages @angular/service-worker",
```
I found that the output is not too verbose and is useful for the time being to see what is going on. We can remove it later if it becomes annoying.

petebacondarwin commented on aio/content/examples/aot-compiler/tsconfig-aot.json:
```diff
@@ -7,7 +7,6 @@
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
     "lib": ["es2015", "dom"],
-    "noImplicitAny": true,
```
Just helps to keep the example tighter. Not really needed.

petebacondarwin replied:
Probably removed it while fishing around trying to get this example to work :-)

IgorMinar replied:
can we keep it?

petebacondarwin replied:
OK

petebacondarwin replied:
done

petebacondarwin commented on aio/tools/examples/example-boilerplate.js:
```diff
@@ -99,13 +96,9 @@ class ExampleBoilerPlate {
-    } else {
-      ngPackagesInstaller.restorePackages(basePath);
-    }
+    const tool = 'node tools/ng-packages-installer';
```
Good point, I will resolve the path, just to be sure.

I used node because I wanted to limit the use of the programmatic API from outside in case it changed. The command line API provides what we need and so I thought it was safer to just use that consistently.

petebacondarwin replied:
What I would like to do is remove this method altogether, since installing different packages is actually different to adding the boilerplate. Once the boilerplate is installed you can switch back and forth between local and npm packages without having to re-run add boilerplate.

petebacondarwin replied:
I have now removed it in e9c72234490001c1dddfc2f3bf5d91aa6ba9b14a

petebacondarwin commented on aio/tools/ng-packages-installer/index.js:
```diff
@@ -0,0 +1,220 @@
+const chalk = require('chalk');
```
OMG!

petebacondarwin replied:
done

petebacondarwin commented on aio/tools/ng-packages-installer/index.js:
```diff
@@ -0,0 +1,220 @@
+const path = require('canonical-path');
+const shelljs = require('shelljs');
+const yargs = require('yargs');
+
```
'cos I'm a cowboy?

petebacondarwin commented on aio/tools/ng-packages-installer/index.js:
```diff
@@ -0,0 +1,220 @@
+const PACKAGE_JSON = 'package.json';
+const LOCKFILE = 'yarn.lock';
+const LOCAL_MARKER_PATH = 'node_modules/_local_.json';
+const PACKAGE_JSON_REGEX = new RegExp(`^[^/]+/${PACKAGE_JSON}$`);
```
Oops!

petebacondarwin commented on aio/tools/ng-packages-installer/index.js:
```diff
@@ -0,0 +1,220 @@
+
+class NgPackagesInstaller {
+
+  constructor(projectDir, options = {}) {
```
What are docs?

petebacondarwin replied:
https://github.com/angular/angular/issues/19573

gkalpak replied:
I mean inline jsdoc-esque docs (for code maintainers - not authors). I could use some for the methods too :grin: Especially, `@param` ones.

petebacondarwin commented on aio/tools/ng-packages-installer/index.js:
```diff
@@ -0,0 +1,220 @@
+
+  _printWarning() {
+    const relativeScriptPath = path.relative('.', __filename.replace(/\.js$/, ''));
+    const absoluteProjectDir = path.resolve(this.projectDir);
```
I guess you preferred the relative path?
Trouble was that I found the message confusing when the relative path was `.`. 
It read like it was a full stop and we had forgotten to write the path.
I guess we could put it in quotes but this way it is unambiguous.

gkalpak replied:
Yeah, it's fine 😄

petebacondarwin committed:
ci(aio): use custom package.json to run with local distributables

Closes #19388

petebacondarwin committed:
build(aio): support ignoring dist packages in "local" mode

gkalpak commented on aio/package.json:
```diff
@@ -6,6 +6,9 @@
   "license": "MIT",
   "scripts": {
+    "aio-use-local": "node tools/ng-packages-installer overwrite . --debug --ignore-packages @angular/service-worker",
+    "aio-use-npm": "node tools/ng-packages-installer restore .",
```
Oh, `setup` and `setup-local` should use these scripts too.

petebacondarwin replied:
done

petebacondarwin committed:
build(aio): provide cleaner scripts for local build switching

petebacondarwin committed:
docs(aio): fix aot-compiler example to work with Angular v5

petebacondarwin committed:
build(aio): example-boilerplate is no longer responsible for yarn install

The tooling for boilerplate was also running `yarn install` on the examples'
shared folder. But since this is handled by `ng-packages-installer` this
commit refactors the tools so that the boilerplate no longer does this
anymore.

mary-poppins commented:
You can preview e9c7223 at https://pr19511-e9c7223.ngbuilds.io/.
